lib/src/deprecations.ts

```diff
@@ -25,53 +25,103 @@ export function getDeprecationIds(
   });
 }
 
+/**
+ * Map between active compilations and the deprecation options they use.
+ *
+ * This is used to determine which options to use when handling host-side
+ * deprecation warnings that aren't explicitly tied to a particular compilation.
+ */
+export const activeDeprecationOptions: Map<Object, DeprecationOptions> =
+  new Map();
+
 /**
  * Shorthand for the subset of options related to deprecations.
  */
-export type DeprecationOptions = {
-  fatalDeprecations?: (DeprecationOrId | Version)[];
-  futureDeprecations?: DeprecationOrId[];
-  silenceDeprecations?: DeprecationOrId[];
-};
+export type DeprecationOptions = Pick<
+  Options<'sync'>,
+  'fatalDeprecations' | 'futureDeprecations' | 'silenceDeprecations'
+>;
 
 /**
  * Handles a host-side deprecation warning, either emitting a warning, throwing
- * and error, or doing nothing depending on the deprecation options used.
+ * an error, or doing nothing depending on the deprecation options used.
+ *
+ * If no specific deprecation options are passed here, then options will be
+ * determined based on the options of the active compilations.
  */
 export function warnForHostSideDeprecation(
   message: string,
   deprecation: Deprecation,
   options?: DeprecationOptions
 ): void {
+  if (options) throw Error(message);
   if (
     deprecation.status === 'future' &&
-    !getDeprecationIds(options?.futureDeprecations ?? []).includes(
-      deprecation.id
-    )
+    !isEnabledFuture(deprecation, options)
   ) {
     return;
   }
   const fullMessage = `Deprecation [${deprecation.id}]: ${message}`;
   if (isFatal(deprecation, options)) {
     throw Error(fullMessage);
   }
-  if (
-    !getDeprecationIds(options?.silenceDeprecations ?? []).includes(
-      deprecation.id
-    )
-  ) {
+  if (!isSilent(deprecation, options)) {
     console.warn(fullMessage);
   }
 }
 
+/**
+ * Checks whether the given deprecation is included in the given list of silent
+ * deprecations or is silenced by at least one active compilation.
+ */
+function isSilent(
+  deprecation: Deprecation,
+  options?: DeprecationOptions
+): boolean {
+  if (!options) {
+    for (const potentialOptions of activeDeprecationOptions.values()) {
+      if (isSilent(deprecation, potentialOptions)) return true;
+    }
+    return false;
+  }
+  return getDeprecationIds(options?.silenceDeprecations ?? []).includes(
```

Comment:
```suggestion
  return getDeprecationIds(options.silenceDeprecations ?? []).includes(
```

There's no way for `options` to be undefined here. Also below.

Reply:
Fixed